android: Set SDK/API level via version-suffxed `--target` triple

MarijnS95 opened the pull request:
Fixes https://github.com/Traverse-Research/evolve/issues/1058. Our version of https://github.com/rust-mobile/xbuild/pull/209, see that PR for all details.

jsclary committed:
Support the x86_64-linux-android build target.

* Initial support for x86_64-linux-android targets.

* Improve target selection for clang link phase.

* Only add clang target if it differs from the host.

MarijnS95 committed:
android: Set SDK/API level via version-suffxed `--target` triple

We haven't set the SDK/API level via the `__ANDROID_API__` define for
a very long time and so far got away with it.  However, while debugging
why `backtrace` (and by extension Rust `std` which reuses that crate)
wasn't generating symbolicated stacktraces in `panic_log`, and why
`findshlibs` wasn't providing the list of loaded libraries to `sentry`,
we found that both rely on expanding the `__ANDROID_API__` define
via compiling a small C file via `cc` to make the code conditional on
SDK/API >= 21:

https://github.com/gimli-rs/findshlibs/pull/65
https://github.com/rust-lang/backtrace-rs/pull/415

(It would have been lovely if these crates emitted a `cargo:warning`
when the define wasn't set at all, indicating an "incomplete"
cross-compiler setup, and/or looked at the runtime Android API version
via something like https://github.com/rust-mobile/ndk/pull/479.)

Note that `backtrace 0.3.74` / Rust 1.82
(https://github.com/rust-lang/rust/commit/0763a3a) no longer rely on
this because Rust 1.82 bumped the minimum SDK/API level to 21:
https://github.com/rust-lang/backtrace-rs/pull/656

We could set this define directly, or rely on `clang` to set it
for us by appending the SDK/API level to the target triple, of
the form `<arch>-linux-android<sdk level>`.  The latter is more
common. Keep in mind that the `cc` crate adds an unversioned
`--target=<arch>-linux-android` to the command line arguments as well,
but clang seems to deduplicate them (or look at the latter `--target`
which contains our version).

Note that this effectively reverts
https://github.com/rust-mobile/xbuild/pull/127/commits/32efed6 because
we must now always pass the SDK level via the triple again, even if the
host also happens to be Android with the same architecture.

MarijnS95 commented:
This seems to work on small test-apps, but fails on our main app with:

```
01-07 17:39:59.432  1141  1141 E AndroidRuntime: java.lang.UnsatisfiedLinkError: dlopen failed: TLS symbol "(null)" in dlopened "/data/app/~~cHL8LBKb7eumhNxrtXw63g==/nl.traverse_research.evolve-u9FL5VS6IhAJMmQ9_7K5kg==/lib/arm64/libevolve_android.so" referenced from "/data/app/~~cHL8LBKb7eumhNxrtXw63g==/nl.traverse_research.evolve-u9FL5VS6IhAJMmQ9_7K5kg==/lib/arm64/libevolve_android.so" using IE access model
```

I did not run into that before testing, perhaps because I upgraded the `backtrace` crate before? Testing ongoing.

MarijnS95 commented:
A lot of chasing later, I got this figured out. Wasn't going to settle for a `-D__ANDROID_API__={min_sdk_ver}` hack.

I couldn't initially reproduce this with `cargo-apk` because I had a too-old NDK installed (`26.3.11579264`). After upgrading to `27.2`, the issue reproduces there as well.

Searching for it, it's supposedly an intended change where ELF TLS is automatically enabled for SDK level `>=29` in upstream LLVM and also in NDK r26b (I was supposedly on r26a): https://github.com/android/ndk/issues/1679. In `xbuild`, we weren't telling `clang` what min API version to compile for.

Via https://github.com/golang/go/issues/69109#issuecomment-2408792678 we also see that the `STATIC_TLS` flag is added to our binary. Turns out that's not because of a compiler flag (or lack thereof), but because the `rpmalloc` crate sets a static TLS model which results in this error: https://github.com/mjansson/rpmalloc/issues/332 / https://github.com/mjansson/rpmalloc/pull/347.

I don't know if we can opt-out of this with a compiler flag, but there are a few things we can do:
1. `git` reference the linked `rpmalloc` PR to fix this;
2. Disable `rpmalloc` for Android (we also disabled it for MacOS due to similar strange issues);
3. Keep not telling `clang` about our API version, and create a bigger gap with the Android ecosystem by hacking it with `-D__ANDROID_API__={min_sdk_ver}` instead :vomiting_face:

maxded commented:
I would be fine with the git reference fix so that we can move on and actually get further into our certification process. Would be good to also share this information with the xbuild issue to get some outside eyes on it (if anyone even does that 👀 ).

> Disable rpmalloc for Android (we also disabled it for MacOS due to similar strange issues);

I'm unsure what kind of impact this would have but could be an option as well, I'll leave the decision up to you.

MarijnS95 commented:
@maxded thanks, that's the way I'm leaning too. Using that commit is quite a divergence so I might just cherry-pick it to an `rpmalloc` fork.

MarijnS95 commented:
@maxded note that this repository is public (it is our fork of `xbuild`) and I can easily link or copy-paste the issue may anyone else observe it.

I'm only opening the PR here _and on the upstream side_ because we have quite a divergent branch where none of the new features have been written/PRd against upstream. I would have to go through everything that was pushed and PR it upstream so that we can get rid of it, or at the very least rebase when all these fixes land upstream.

MarijnS95 committed:
CI: Drop broken template building

We don't use the template and its `wry` feature in this fork.
Remove it now that `ubuntu-latest` no longer has the necessary
packages which requires a tedious and complicated upgrade
of the `wry` "nonsense".  This is attempted upstream at:
https://togithub.com/rust-mobile/xbuild/pull/210
